lib/core/src/display/workspace.rs

```diff
@@ -19,145 +31,218 @@ pub enum Error {
+        let sub_logger    = logger.sub("mesh_registry");
+        let mesh_registry = MeshRegistry::new(sub_logger, on_change);
+        let shape         = default();
+        let listeners     = Self::init_listeners(&canvas,&shape,&shape_dirty);
+        let mesh_registry_dirty = dirty_flag;
+        let this = Self {canvas,context,mesh_registry,mesh_registry_dirty
+                        ,shape,shape_dirty,logger,listeners};
+        Ok(this)
     }
-
-    pub fn resize(&self, width: i32, height: i32) {
+    /// Initialize all listeners and attach them to DOM elements.
+    fn init_listeners
+    ( canvas : &web_sys::HtmlCanvasElement
+    , shape  : &Rc<RefCell<WorkspaceShape>>
+    , dirty  : &ShapeDirty<OnDirty>
+    ) -> Listeners {
+        let shape = shape.clone();
+        let dirty = dirty.clone();
+        let on_resize = Closure::new(move |width, height| {
+            *shape.borrow_mut() = WorkspaceShape {width,height};
+            dirty.set();
+        });
+        let resize = ResizeObserver::new(canvas,on_resize);
+        Listeners {resize}
+    }
+    /// Build new instance with the provided builder object.
+    pub fn build<Name:Into<String>> (name:Name) -> WorkspaceBuilder {
+        let name = name.into();
+        WorkspaceBuilder {name}
+    }
+    /// Create a new mesh instance.
+    pub fn new_mesh(&mut self) -> MeshID {
+        self.mesh_registry.new_mesh()
+    }
+    /// Resize the underlying canvas. This function should rather not be called
+    /// directly. If you want to change the canvas size, modify the `shape` and
+    /// set the dirty flag.
+    fn resize_canvas(&self, shape:&WorkspaceShape) {
+        let width  = shape.width;
+        let height = shape.height;
         self.logger.group(fmt!("Resized to {}px x {}px.", width, height), || {
             self.canvas.set_attribute("width", &width.to_string()).unwrap();
             self.canvas.set_attribute("height", &height.to_string()).unwrap();
             self.context.viewport(0, 0, width, height);
-            self.dirty.set();
         });
     }
+
+    // TODO TODO  TODO  TODO  TODO  TODO  TODO  TODO  TODO  TODO  TODO  TODO 
+    // THIS FUNCTION WILL BE REFACTORED IN THE NEAR FUTURE. IT IS A ROUGH
+    // MOCK NOW. DO NOT REVIEW IT.
+    /// Check dirty flags and update the state accordingly.
+    pub fn update(&mut self) {
+        group!(self.logger, "Updating.", {
+            if self.shape_dirty.check() {
+                self.resize_canvas(&self.shape.borrow());
+                self.shape_dirty.unset();
+            }
+            if self.mesh_registry_dirty.check() {
+                self.mesh_registry.update();
+                self.mesh_registry_dirty.unset();
+            }
+        
+            let vert_shader = webgl::compile_shader(
+                &self.context,
+                webgl::Context::VERTEX_SHADER,
+                r#"
+    attribute vec4 position;
+    void main() {
+        gl_Position = position;
+    }
+"#,
```

Comment:
What if we use `include_str!` here?

Like webgl::compile_shader(&self.context, webgl::Context::VERTEX_SHADER, include_str!("vertex_shader.glsl"))

It will be easier to edit the code in `vertex_shader.glsl` with code highlighting.

Reply:
That would definitely be better, however, we should not have GLSL code pasted to our code. As my comment in the header of this function tells: `THIS FUNCTION WILL BE REFACTORED IN THE NEAR FUTURE. IT IS A ROUGH MOCK NOW. DO NOT REVIEW IT.`. I would never allow such ugly function written in such a bad style as this one in final code - this is something I will be refactoring out and heavily changing in next PR. There would definitely not be GLSL anymore soon :)